### PR TITLE
Custom cop changes now bust the cache

### LINKED ISSUE #2886: Cache does not depend on content of Cops
If you run `rubocop` on a file, then change a cop to do something different, then re-run `rubocop` it will just return the previous result. `-C false` fixes it, but that is pretty bizarre and unexpected for cop authors. Even worse, cop changes in CI will continue running old versions until the files are touched.

I don't know what the current cache key is, but can we include a SHA of the cops that run too in there?

### PR DESCRIPTION
**Pull request by ptarjan**
Fixes #2886

Before submitting a PR make sure the following are checked:

* [x] Wrote [good commit messages][1].
* [x] Used the same coding conventions as the rest of the project.
* [x] Feature branch is up-to-date with `master` (if not - rebase it)
* [x] Squashed related commits together.
* [ ] Added tests.
* [x] Added an entry to the [Changelog](CHANGELOG.md) if the new code introduces user-observable changes. See [changelog entry format](../CONTRIBUTING.md#changelog-entry-format).
* [x] All tests are passing.
* [x] The new code doesn't generate RuboCop offenses.
* [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.

[1]: http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

### CODE REVIEW

File: lib/rubocop/result_cache.rb [hunk 1]
```diff
@@ -128,7 +128,8 @@ def rubocop_checksum
         begin
           lib_root = File.join(File.dirname(__FILE__), '..')
           bin_root = File.join(lib_root, '..', 'bin')
-          source = Find.find(lib_root, bin_root).sort.map do |path|
+          source_files = $LOADED_FEATURES + Find.find(lib_root, bin_root).to_a
```
Comment (bbatsov) [inline lib/rubocop/result_cache.rb]: I wonder if this is the best way to solve this. After all, most of the files here probably won't be custom cops.

Comment (jonas054) [inline lib/rubocop/result_cache.rb]: It's a very simple way of finding all the source files that can affect RuboCop's behavior, which is why I suggested it in #2886. I don't think it affects the start-up time severely in the normal case (a reasonably performant file system). The files under `lib` will be included in `$LOADED_FEATURES` BTW, so you can remove `lib_root`.

Comment (bbatsov) [inline lib/rubocop/result_cache.rb]: At the very least we should have a comment here explaining what this line of code does. Even I had to think a bit about it when I saw it.

File: CHANGELOG.md [hunk 2]
```diff
@@ -18,6 +18,7 @@
 * [#2975](https://github.com/bbatsov/rubocop/issues/2975): Make comment indentation before `)` consistent with comment indentation before `}` or `]`. ([@jonas054][])
 * [#3010](https://github.com/bbatsov/rubocop/issues/3010): Fix double reporting/correction of spaces after ternary operator colons (now only reported by `Style/SpaceAroundOperators`, and not `Style/SpaceAfterColon` too). ([@owst][])
 * [#3006](https://github.com/bbatsov/rubocop/issues/3006): Register an offense for calling `merge!` on a method on a variable inside `each_with_object` in `Performance/RedundantMerge`. ([@lumeet][], [@rrosenblum][])
+* [#2886](https://github.com/bbatsov/rubocop/issues/2886): Custom Cops changes now bust the cache. ([@ptarjan][])
```
Comment (bbatsov) [inline CHANGELOG.md]: Cops -> cops

**Comment by bbatsov** [comment]
I think the commit message is a bit unclear right now (the changelog entry seems better formulated). Don't remember to capitalize the first word in the commit message.

File: lib/rubocop/result_cache.rb (hunk 1 above)
Comment (ptarjan) [inline lib/rubocop/result_cache.rb]: I could record the things we `require`d but this is safer. I didn't profile it, but if this turns out to be the lowest hanging startup fruit we can change it.

**Comment by ptarjan** [comment]
@bbatsov updated

**Pull request merged**